### PR TITLE
KotlinMentionsProcessor enhancement

### PR DESCRIPTION
**Pull request by sleshJdev**
A few improvements along with refactoring:

1. Enhanced kotlin regex to handle typoes(pseudo letters, register, duplicated symbols etc)
2. Introduced random delay to get the bot to behave a bit more natural 
3. Upgraded gradlew to support JUnit's parameterized test cases
4. Updated sticker message wording

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.0.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.5.1-bin.zip
```
Comment (sleshJdev) [inline gradle/wrapper/gradle-wrapper.properties:3]: Please, see the related gradle's [issue](https://github.com/gradle/gradle/issues/5975#issuecomment-591670916)

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt [hunk 2]
```diff
@@ -14,28 +14,39 @@ import com.github.insanusmokrassar.TelegramBotAPI.types.update.MessageUpdate
 import com.github.insanusmokrassar.TelegramBotAPI.types.update.abstracts.Update
 import java.time.Duration
 import java.time.Instant
+import kotlin.random.Random
 
 class KotlinMentionsProcessor(
         private val bot: RequestsExecutor,
-        private val kotlinMentionsDAO: KotlinMentionsDAO
+        private val kotlinMentionsDAO: KotlinMentionsDAO,
```
Comment (sleshJdev) [inline src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:21]: Just refactoring to make it a bit cuter and flexible. It's not quite related to the PR's goal but I decide to make it as well to reduce the number of PRs

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt [hunk 3]
```diff
@@ -14,28 +14,39 @@ import com.github.insanusmokrassar.TelegramBotAPI.types.update.MessageUpdate
...
+        private val stickerFileId: String = zeroDaysWithoutKotlinStickerFileId,
+        private val hasPassedEnoughTimeSincePreviousReply: (Duration) -> Boolean = ::hasPassedEnoughTimeSincePreviousReply,
+        private val containsMatchIn: (String) -> Boolean = kotlinRegex::containsMatchIn,
+        private val composeStickerMessage: (Duration) -> String = ::composeStickerMessage
 ) : UpdateProcessor {
     companion object {
         const val zeroDaysWithoutKotlinStickerFileId = "CAACAgIAAxkBAAIBsF8V0dPb6EesBKSujFFOx_URfhSdAAJAAQACqSImBOs5DmSNtKlmGgQ"
-        val kotlinRegex = "(котлин|kotlin)".toRegex(setOf(RegexOption.IGNORE_CASE, RegexOption.MULTILINE))
-        val requiredDelayBetweenReplies: Duration? = Duration.ofHours(1)
+        val kotlinRegex = "([kк]+[оo0]+[тt]+[лl]+[ие1ie]+[нn]+)".toRegex(setOf(RegexOption.IGNORE_CASE, RegexOption.MULTILINE))
```
Comment (sleshJdev) [inline src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:29]: Enhanced regex

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt [hunk 4]
```diff
@@ -14,28 +14,39 @@ import com.github.insanusmokrassar.TelegramBotAPI.types.update.MessageUpdate
...
+        private val minRequiredDelayBetweenReplies: Duration = Duration.ofMinutes(5)!!
```
Comment (sleshJdev) [inline src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:30]: Random delay in the range

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt [hunk 5]
```diff
@@ -14,28 +14,39 @@ import com.github.insanusmokrassar.TelegramBotAPI.types.update.MessageUpdate
...
+        private val maxRequiredDelayBetweenReplies: Duration = Duration.ofHours(1)!!
 
-        fun composeWithoutMentionDurationMessage(duration: Duration): String {
-            return "We have been existing %02dd:%02dh:%02dm:%02ds without mentioning"
-                    .format(duration.toDaysPart(), duration.toHoursPart(), duration.toMinutesPart(), duration.toSecondsPart())
-        }
+        fun composeStickerMessage(duration: Duration): String =
+                "Passed %02dd:%02dh:%02dm:%02ds without an incident".format(
```
Comment (sleshJdev) [inline src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:34]: Updated sticker message wording

File: src/test/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessorTest.kt [hunk 6]
```diff
@@ -93,46 +95,32 @@ class KotlinMentionsProcessorTest {
...
+        "А если kкоo0тtлlие1ieнn?", "а с котлином прокатывает)", "Jsf c kotlin dsl", "Kotlin с иммутабельными датаклассами",
+        """
             лююдиии
             мб кто-то помнит твит какого-то чувак, где он говорит что kotlin разработчики были бы намного эффективнее, 
             если бы не тратили кучу времени доказывая всем вокруг что котлин крут
             никак найти не могу...
-        """.trimIndent())
-    }
+        """])
+    fun `test mentioning fuzzy kotlin then should be send`(kotlinTypo: String) = testStickerWasSent(kotlinTypo)
```
Comment (sleshJdev) [inline src/test/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessorTest.kt:123]: Collapsed the all kotlin-word-speling tests into parameterized one

**Comment by madhead** [comment]
@sleshJdev, hey, what about introducting mentions statistics: who mentioned Kotlin and how many times? And then making it accessible via a command, like `/kotlin-mentions`?

**Comment by sleshJdev** [comment]
> @sleshJdev, hey, what about introducting mentions statistics: who mentioned Kotlin and how many times? And then making it accessible via a command, like `/kotlin-mentions`?

Hey

Why not, but it's matter of separate pr

**Comment by madhead** [comment]
@sleshJdev would you be kind enough to make it? :)

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt [hunk 7]
```diff
@@ -14,28 +14,39 @@ import com.github.insanusmokrassar.TelegramBotAPI.types.update.MessageUpdate
 import com.github.insanusmokrassar.TelegramBotAPI.types.update.abstracts.Update
 import java.time.Duration
 import java.time.Instant
+import kotlin.random.Random
 
 class KotlinMentionsProcessor(
         private val bot: RequestsExecutor,
-        private val kotlinMentionsDAO: KotlinMentionsDAO
+        private val kotlinMentionsDAO: KotlinMentionsDAO,
+        private val stickerFileId: String = zeroDaysWithoutKotlinStickerFileId,
```
Comment (madhead) [inline src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:22]: What is the purpose of parametrizing the processor with it's own companion's field?

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt [hunk 8]
```diff
@@ -14,28 +14,39 @@ import com.github.insanusmokrassar.TelegramBotAPI.types.update.MessageUpdate
...
+        private val hasPassedEnoughTimeSincePreviousReply: (Duration) -> Boolean = ::hasPassedEnoughTimeSincePreviousReply,
```
Comment (madhead) [inline src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:23]: Again, why parametrizing the class with it's companion's method?

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt [hunk 9]
```diff
@@ -14,28 +14,39 @@ import com.github.insanusmokrassar.TelegramBotAPI.types.update.MessageUpdate
...
-        fun composeWithoutMentionDurationMessage(duration: Duration): String {
-            return "We have been existing %02dd:%02dh:%02dm:%02ds without mentioning"
-                    .format(duration.toDaysPart(), duration.toHoursPart(), duration.toMinutesPart(), duration.toSecondsPart())
-        }
+        fun composeStickerMessage(duration: Duration): String =
+                "Passed %02dd:%02dh:%02dm:%02ds without an incident".format(
+                        duration.toDaysPart(), duration.toHoursPart(),
+                        duration.toMinutesPart(), duration.toSecondsPart())
+
+        fun hasPassedEnoughTimeSincePreviousReply(duration: Duration): Boolean =
```
Comment (madhead) [inline src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:38]: Why "reply" and not "mention"?

**Comment by sleshJdev** [comment]
> @sleshJdev would you be kind enough to make it? :)

Yeah, I am fine to accomplishe it in a separate pr

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt (hunk 9 above)
Comment (sleshJdev) [reply src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt]: No good reason just kinda inattention. Fixed f8ce454ef65abe96a96eb135620eb1585a10db4b

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt (hunk 8 above)
Comment (sleshJdev) [reply src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt]: To keep the constructor signature clean. By putting too much logic into the method signature it needs more mental effort to read it. I prefer to encapsulate complex logic(not a short oneliner) into a dedicated method and refer it

File: src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt (hunk 7 above)
Comment (sleshJdev) [reply src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:22]: It's kind of default and could be overridden if need. Overall this refactoring doesn't bring a lot of value but it looks better from the code design perspective - it's the main reason why I made it.

Comment (sleshJdev) [reply src/main/kotlin/by/jprof/telegram/opinions/processors/KotlinMentionsProcessor.kt:22]: zeroDaysWithoutKotlinStickerFileId brings some hint to you by its name, using a literal value would confuse

**Comment by madhead** [comment]
@sleshJdev, deployed to production 🎉

Thank you for the contribution!

**Pull request merged**